src/main/java/com/google/devtools/build/lib/actions/ActionCacheChecker.java

```diff
@@ -287,6 +359,12 @@ public Token getTokenIfNeedToExecute(
       }
     }
     ActionCache.Entry entry = getCacheEntry(action);
+
+    // load remote metadata from action cache
+    if (entry != null) {
+      loadRemoteOutputMetadataIfMissing(action, entry, metadataHandler);
```

Comment:
I wonder if we shouldn't inject the outputs only if we get an action cache hit (after mustExecute). I understand that outputStore is currently used to get the metadata, but we could change `getMetadataMaybe` to look for the metadata in the entry itself or in some intermediary which has the loaded, but not yet injected files and trees.

My worry here would be if a remote executing action for some reason does not produce an output it should, but got it injected from the output store to begin with -- can we hold onto a stale output this way?

Reply:
OutputStore is [cleared](https://github.com/bazelbuild/bazel/blob/aad74a3646d30d0b6a440a566f06aa264697676d/src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java#L467) right [before](https://github.com/bazelbuild/bazel/blob/aad74a3646d30d0b6a440a566f06aa264697676d/src/main/java/com/google/devtools/build/lib/skyframe/ActionExecutionFunction.java#L806) action execution so we should be safe in that case.

That said, I updated the code to follow your suggestion in order to decouple from that assumption of the current `ActionMetadataHandler`.